google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/changestream/ReadChangeStreamRetryTest.java

```diff
@@ -0,0 +1,494 @@
+    service.expectations.add(
+        RpcExpectation.create().expectInitialRequest().respondWith(closeStreamResponse));
+    List<ChangeStreamRecord> actualResults = getResults();
+    Assert.assertEquals(actualResults.size(), 1);
+    Assert.assertTrue(actualResults.get(0) instanceof CloseStream);
+  }
+
+  // [{UNAVAILABLE}, {DataChange with done==true}] -> [{(ChangeStreamRecord) ChangeStreamMutation}]
+  @Test
+  public void singleCompleteDataChangeImmediateRetryTest() {
+    // DataChange
+    ReadChangeStreamResponse dataChangeResponse =
+        ReadChangeStreamResponse.newBuilder().setDataChange(createDataChange(true)).build();
+    service.expectations.add(
+        RpcExpectation.create().expectInitialRequest().respondWithStatus(Code.UNAVAILABLE));
+    // Resume with the exact same request.
+    service.expectations.add(
+        RpcExpectation.create().expectInitialRequest().respondWith(dataChangeResponse));
+    List<ChangeStreamRecord> actualResults = getResults();
+    Assert.assertEquals(actualResults.size(), 1);
+    Assert.assertTrue(actualResults.get(0) instanceof ChangeStreamMutation);
+  }
+
+  // [{ReadChangeStreamResponse.Heartbeat}, {UNAVAILABLE}] -> Resume with token from heartbeat.
+  @Test
+  public void errorAfterHeartbeatShouldResumeWithTokenTest() {
+    StreamContinuationToken streamContinuationToken =
+        createStreamContinuationToken(HEARTBEAT_TOKEN);
+    ReadChangeStreamResponse heartbeatResponse =
+        ReadChangeStreamResponse.newBuilder()
+            .setHeartbeat(createHeartbeat(streamContinuationToken))
+            .build();
+    service.expectations.add(
+        RpcExpectation.create()
+            .expectInitialRequest()
+            .respondWith(heartbeatResponse)
+            .respondWithStatus(Code.UNAVAILABLE));
+    // Resume the request with the token from the Heartbeat. `startTime` is cleared.
+    // We don't care about the response here so just do expectRequest.
+    service.expectations.add(
+        RpcExpectation.create()
+            .expectRequest(
+                StreamContinuationTokens.newBuilder().addTokens(streamContinuationToken).build()));
+    List<ChangeStreamRecord> actualResults = getResults();
+    // This is the Heartbeat we get before UNAVAILABLE.
+    Assert.assertEquals(actualResults.size(), 1);
+    Assert.assertTrue(actualResults.get(0) instanceof Heartbeat);
+  }
+
+  // [{ReadChangeStreamResponse.CloseStream}, {UNAVAILABLE}] -> Resume with original request.
+  @Test
+  public void errorAfterSingleCloseStreamShouldResumeWithOriginalRequestTest() {
+    // CloseStream.
+    ReadChangeStreamResponse closeStreamResponse =
+        ReadChangeStreamResponse.newBuilder().setCloseStream(createCloseStream()).build();
+    service.expectations.add(
+        RpcExpectation.create()
+            .expectInitialRequest()
+            .respondWith(closeStreamResponse)
+            .respondWithStatus(Code.UNAVAILABLE));
+    // Resume the request with the original request.
```

Comment:
Should we resume the request in this case if we already see close stream response? Will we get same response if we send the original request?